src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -825,31 +834,80 @@ impl Coordinator {
+            user: ctx.session().user().clone(),
+            conn_id: ctx.session().conn_id().clone(),
+            sender: Some(tx),
+            cluster_id,
+            depends_on: validity.dependency_ids.clone(),
+            dropping: false,
+        };
+        // Add metadata for the new COPY TO.
+        self.add_active_copy_to(sink_id, active_copy_to);
+
         let (df_desc, df_meta) = global_lir_plan.unapply();
 
-        self.emit_optimizer_notices(&*session, &df_meta.optimizer_notices);
+        self.emit_optimizer_notices(ctx.session(), &df_meta.optimizer_notices);
 
-        // Ship the dataflow
-        self.controller
-            .active_compute()
-            .create_dataflow(optimizer.cluster_id(), df_desc)
-            .unwrap_or_terminate("cannot fail to create dataflows");
-
-        // TODO(mouli): implement!
-        Err(AdapterError::Internal(format!(
-            "COPY TO '{}' is not yet implemented",
-            optimizer.copy_to_uri()
-        )))
+        // Ship dataflow.
+        self.ship_dataflow(df_desc, cluster_id).await;
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        // Awaiting the response in a separate task to unblock
+        // further processing.
+        mz_ore::task::spawn(|| "copy_to await response", async move {
+            let response = match rx.await {
+                Ok(res) => Some(res),
+                Err(error) => {
+                    warn!("error receiving response back from sink: {:?}", error);
+                    None
+                }
+            };
+            let _ = internal_cmd_tx.send(Message::PeekStageReady {
+                ctx,
+                otel_ctx: root_otel_ctx,
+                stage: PeekStage::CopyToFinish(PeekStageCopyToFinish { validity, response }),
+            });
+        });
+    }
+
+    fn peek_stage_copy_to_finish(
+        &mut self,
+        PeekStageCopyToFinish {
+            validity: _,
+            response,
+        }: PeekStageCopyToFinish,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        let response = response.ok_or_else(|| {
+            AdapterError::Unstructured(anyhow!("unexpected error while fetching response"))
+        })?;
+        match response {
+            CopyToResponse::RowCount(n) => {
+                Ok(ExecuteResponse::Copied(n.try_into().expect("u64 to usize")))
+            }
+            CopyToResponse::Error(error) => Err(AdapterError::Unstructured(anyhow!(error))),
+            // We do not expect to receive a `Dropped` message here. Say, when one replica
+            // has finished the copying, we drop the compute sinks from the adapter
+            // after receiving one response. The unfinished sinks will emit a `Dropped` message, but
+            // these will not be sent to the adapter because a message was sent previously (the controller
+            // will only send one CopyToResponse for one corresponding COPY TO)
+            // TODO (mouli): should we panic! instead?
```

Comment:
I'd recommend using types to enforce this statically if possible. It looks to me like the layer of the stack that "swallows" `CopyToResponse::Dropped` (I think the controller in `compute-client`) could convert the underlying `CopyToResponse` to a different type that doesn't have a `Dropped` variant.

On a quick skim, it looks like `ComputeResponse::CopyToResponse(GlobalId, CopyToResponse)` potentially wants to be `ComputeResponse::CopyToResponse(GlobalId, Result<u64, Error>)`?

Reply:
ack, makes sense, updating

Reply:
Updated. Converting to `Result<u64, Error>` in the `compute-client`. 

The `ComputeResponse::CopyToResponse` is still `(GlobalId, CopyToResponse)` because that's what gets aggregated in compute's `PartitionedState`, but the `ComputeControllerResponse::CopyToResponse` and then the `ControllerResponse::CopyToResponse` is using `Result`.